[quant][fx2trt] Add lowering support for reference linear/conv modules

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64787
* #64857
* __->__ #64368

Summary:

Test Plan:
python torch/fx/experimental/fx2trt/example/quantized_resnet_test.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30708738](https://our.internmc.facebook.com/intern/diff/D30708738)

jerryzh168 committed:
[quant][fx2trt] Add lowering support for reference linear/conv modules

Summary:

Test Plan:
python torch/fx/experimental/fx2trt/example/quantized_resnet_test.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx2trt] Add lowering support for reference linear/conv modules"


Summary:

Test Plan:
python torch/fx/experimental/fx2trt/example/quantized_resnet_test.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30708738).

jerryzh168 committed:
Update on "[quant][fx2trt] Add lowering support for reference linear/conv modules"


Summary:

Test Plan:
python torch/fx/experimental/fx2trt/example/quantized_resnet_test.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30708738](https://our.internmc.facebook.com/intern/diff/D30708738)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30708738).

jerryzh168 committed:
Update on "[quant][fx2trt] Add lowering support for reference linear/conv modules"


Summary:

Test Plan:
python torch/fx/experimental/fx2trt/example/quantized_resnet_test.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30708738](https://our.internmc.facebook.com/intern/diff/D30708738)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30708738).

jerryzh168 committed:
Update on "[quant][fx2trt] Add lowering support for reference linear/conv modules"


Summary:

Test Plan:
python torch/fx/experimental/fx2trt/example/quantized_resnet_test.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30708738](https://our.internmc.facebook.com/intern/diff/D30708738)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30708738).

jerryzh168 committed:
Update on "[quant][fx2trt] Add lowering support for reference linear/conv modules"


Summary:

Test Plan:
python torch/fx/experimental/fx2trt/example/quantized_resnet_test.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30708738](https://our.internmc.facebook.com/intern/diff/D30708738)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx2trt] Add lowering support for reference linear/conv modules"


Summary:

Test Plan:
python torch/fx/experimental/fx2trt/example/quantized_resnet_test.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30708738](https://our.internmc.facebook.com/intern/diff/D30708738)

[ghstack-poisoned]

842974287 commented on torch/fx/experimental/fx2trt/converters/acc_ops_converters.py:
```diff
@@ -241,16 +241,35 @@ def acc_ops_conv2d(network, target, args, kwargs, name):
+    # for now we'll assume bias is constant Tensor or None,
+    # and bias being ITensor is not supported in TensorRT api
+    # right now
     bias = to_numpy(kwargs["bias"])
```
Let's add a check in `to_numpy()` to fail loud if sees something that's not a `torch.Tensor`.

842974287 commented on torch/fx/experimental/fx2trt/converters/acc_ops_converters.py:
```diff
@@ -241,16 +241,35 @@ def acc_ops_conv2d(network, target, args, kwargs, name):
-        kernel=kernel,
-        bias=bias,
-    )
+    weight = get_trt_tensor(network, kwargs["weight"], f"{name}_weight")
+    weight_shape = tuple(kwargs["weight"].shape)
```
Maybe move these two lines into the if branch below?

842974287 commented on torch/fx/experimental/fx2trt/converters/acc_ops_converters.py:
```diff
@@ -1019,43 +1038,48 @@ def acc_ops_linear(network, target, args, kwargs, name):
 
-    weight = kwargs["weight"]
-
     # For quantization, weight here would be a trt tensor because it goes through
     # quant + dequant. In this case, we need to use matmul + add because fully_connected
     # can't take non-constant weight.
```
we can remove this part of the comments now.

842974287 commented on torch/fx/experimental/fx2trt/converters/acc_ops_converters.py:
```diff
@@ -1307,7 +1331,9 @@ def acc_ops_permute(network, target, args, kwargs, name):
 @tensorrt_converter(acc_ops.quantize_per_tensor)
 def acc_ops_quantize_per_tensor(network, target, args, kwargs, name):
-    input_val = kwargs["input"]
+    # input_val = kwargs["input"]
```
Remove

842974287 reviewed (approved):
Looks great! Please also check internal CI to make sure all tests passed since right now all fx2trt tests lives in internal CI.

jerryzh168 committed:
Update on "[quant][fx2trt] Add lowering support for reference linear/conv modules"


Summary:

Test Plan:
python torch/fx/experimental/fx2trt/example/quantized_resnet_test.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30708738](https://our.internmc.facebook.com/intern/diff/D30708738)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30708738).

jerryzh168 committed:
Update on "[quant][fx2trt] Add lowering support for reference linear/conv modules"


Summary:

Test Plan:
python torch/fx/experimental/fx2trt/example/quantized_resnet_test.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30708738](https://our.internmc.facebook.com/intern/diff/D30708738)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx2trt] Add lowering support for reference linear/conv modules"


Summary:

Test Plan:
python torch/fx/experimental/fx2trt/example/quantized_resnet_test.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30708738](https://our.internmc.facebook.com/intern/diff/D30708738)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30708738).

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30708738).

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30708738).

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30708738).
